src/infra/task-system/postgres/src/task_system_event_store_postgres.rs

```diff
@@ -0,0 +1,236 @@
+            task_event_id: i64,
+        }
+
+        let rows = sqlx::query_as!(
+            ResultRow,
+            r#"
+            INSERT INTO task_events (task_id, dataset_id, event_time, event_type, event_payload)
+                SELECT * FROM UNNEST(
+                    $1::INT8[],
+                    $2::TEXT[],
+                    $3::timestamptz[],
+                    $4::TEXT[],
+                    $5::JSONB[]
+                ) RETURNING task_event_id
+            "#,
+            &insert_events.task_ids,
+            &insert_events.dataset_ids as _,
+            &insert_events.event_times,
+            &insert_events.event_types as _,
+            &insert_events.event_payloads
+        )
+        .fetch_all(tr.connection_mut())
+        .await
+        .unwrap();
+
+        if let Some(last_row) = rows.last() {
+            Ok(EventID::new(last_row.task_event_id))
+        } else {
+            Ok(EventID::new(
+                i64::try_from(self.len().await.unwrap()).unwrap(),
+            ))
+        }
+    }
+
+    async fn len(&self) -> Result<usize, InternalError> {
+        let mut tr = self.transaction.lock().await;
+
+        let result = sqlx::query!(
+            r#"
+            SELECT COUNT(task_event_id) from task_events
+            "#,
+        )
+        .fetch_one(tr.connection_mut())
+        .await
+        .int_err()?;
+
+        let count = usize::try_from(result.count.unwrap()).int_err()?;
+        Ok(count)
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl TaskSystemEventStore for TaskSystemEventStorePostgres {
+    /// Generates new unique task identifier
+    async fn new_task_id(&self) -> Result<TaskID, InternalError> {
+        let mut tr = self.transaction.lock().await;
+
+        let result = sqlx::query!(
+            r#"
+            SELECT nextval('task_id_seq') as new_task_id
+            "#
+        )
+        .fetch_one(tr.connection_mut())
+        .await
+        .int_err()?;
+
+        Ok(TaskID::new(result.new_task_id.unwrap()))
+    }
+
+    /// Returns page of the tasks associated with the specified dataset in
+    /// reverse chronological order based on creation time
+    async fn get_tasks_by_dataset(
+        &self,
+        dataset_id: &DatasetID,
+        pagination: TaskPaginationOpts,
+    ) -> TaskIDStream {
+        let mut tr = self.transaction.lock().await;
+        let dataset_id = dataset_id.clone();
+
+        Box::pin(async_stream::stream! {
+            let mut query_stream = sqlx::query!(
+                r#"
+                SELECT task_id FROM
+                (
+                    SELECT DISTINCT ON (task_id) task_id, event_time
+                        FROM task_events
+                        WHERE dataset_id = $1
+                        ORDER  BY task_id, event_time ASC
+                )
+                ORDER BY event_time DESC
+                LIMIT $2 OFFSET $3
```

Comment:
Actually related to the comment below. We could have a dedicated `Task` entity with a separate repository. One of the fields it could hold is a dataset reference.

Reply:
Btw, right now the `dataset_id` field is not null only for `TaskEventCreated`.

Reply:
Simplified